### PR TITLE
add audit enable to role

### PR DESCRIPTION
**Pull request by noy93845**
expose audit enabled flag in athens ui
Signed-off-by: nsegal <noy.segal@yahooinc.com>

### CODE REVIEW

File: ui/src/components/role/AddRole.js [hunk 1]
```diff
@@ -171,6 +176,13 @@ class AddRole extends React.Component {
         };
     }
 
+    toggleAuditEnabled() {
+        this.state.role.auditEnabled = !this.state.role.auditEnabled;
```
Comment (royelbaum) [inline ui/src/components/role/AddRole.js:180]: you shouldn't set the value of this.state.role.auditEnabled it destroy the immutable.
you should extract the value to a different var.
look immer library it should help you.

File: ui/src/components/role/AddRole.js [hunk 2]
```diff
@@ -293,10 +305,15 @@ class AddRole extends React.Component {
         let role = this.state.role;
         role.name = roleName;
         role.reviewEnabled = this.state.reviewEnabled;
+        role.auditEnabled = this.state.role.auditEnabled;
```
Comment (royelbaum) [inline ui/src/components/role/AddRole.js:308]: let role = this.state.role doesn't do deep copy and there for its the same comment as the above.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java [hunk 3]
```diff
@@ -8881,7 +8881,7 @@ void validateRoleMetaAuditEnabledFlag(RoleMeta meta, Role role, Domain domain, f
             if (domain.getAuditEnabled() != Boolean.TRUE) {
                 throw ZMSUtils.requestError("Role cannot be set as audit-enabled if the domain is not audit-enabled", caller);
             }
-            if (role.getAuditEnabled() != Boolean.TRUE && !role.getRoleMembers().isEmpty()) {
+            if (role.getAuditEnabled() != Boolean.TRUE && role.getRoleMembers() != null && !role.getRoleMembers().isEmpty()) {
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java:8884]: please make sure to add a unit test case for this function that checks for the null role members object.

File: ui/src/components/role/AddRole.js (hunk 1 above)
Comment (noy93845) [reply ui/src/components/role/AddRole.js]: Hey @royelbaum thank you for the code review!
I went ahead and integrated immer lib in my code.

File: ui/src/components/role/AddRole.js (hunk 2 above)
Comment (noy93845) [reply ui/src/components/role/AddRole.js]: thanks, see comment above about using immer lib

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java (hunk 3 above)
Comment (noy93845) [reply servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java:8884]: Hey @havetisyan thank you for the heads up!
I went ahead and fixed two issues I found (this time with groups) and added tests for them as well.

File: ui/src/components/header/DomainDetails.js [hunk 4]
```diff
@@ -386,6 +386,14 @@ class DomainDetails extends React.Component {
                         </ValueDiv>
                         <LabelDiv>AWS ACCOUNT ID</LabelDiv>
                     </SectionDiv>
+                    <SectionDiv>
+                        <ValueDiv>
+                            {this.props.domainDetails.gcpProject
+                                ? this.props.domainDetails.gcpProject
+                                : 'N/A'}
+                        </ValueDiv>
+                        <LabelDiv>GCP ACCOUNT ID</LabelDiv>
```
Comment (havetisyan) [inline ui/src/components/header/DomainDetails.js:395]: This should be GCP PROJECT ID (no accounts in GCP)

**Review by havetisyan — CHANGES_REQUESTED** [review]
we should not combine features in the same PR. can we move this out and have a separate PR just for this change please

File: ui/src/components/header/DomainDetails.js (hunk 4 above)
Comment (noy93845) [reply ui/src/components/header/DomainDetails.js]: Fixed! New branch for this change: https://github.com/AthenZ/athenz/pull/2065

**Comment by havetisyan** [comment]
@noy93845 server side changes look fine. please follow up with Roy to review UI changes and once Roy approves the changes, we'll merge this PR.

File: ui/src/components/group/AddGroup.js [hunk 5]
```diff
@@ -317,13 +350,26 @@ class AddGroup extends React.Component {
...
                     </SectionDiv>
                 }
-                <SectionDiv>
+                <SectionDiv title={auditToolTip}>
                     <StyledInputLabel />
                     <StyledIncludedMembersDiv>
                         {members}
                     </StyledIncludedMembersDiv>
                 </SectionDiv>
                 {this.getJustification()}
```
Comment (royelbaum) [inline ui/src/components/group/AddGroup.js:359]: in the function there is no need to use justificationRequired instead use isDomainAuditEnabled and stop passing it from GroupList

File: ui/src/components/role/AddRole.js [hunk 6]
```diff
@@ -152,10 +154,14 @@ class AddRole extends React.Component {
         this.delegateChanged = this.delegateChanged.bind(this);
         this.onSubmit = this.onSubmit.bind(this);
         this.expandSettings = this.expandSettings.bind(this);
+        this.toggleAuditEnabled = this.toggleAuditEnabled.bind(this);
         this.dateUtils = new DateUtils();
 
-        let role = {};
-        role['selfServe'] = false;
+        let role = {
```
Comment (royelbaum) [inline ui/src/components/role/AddRole.js:160]: self serve is not used delete it and add audit enable directly to state it is best practice to be consistent in all components, try to do it like you did in addGroup

File: ui/src/components/settings/SettingTable.js [hunk 7]
```diff
@@ -355,6 +366,30 @@ class SettingTable extends React.Component {
                 />
             );
 
+        let auditEnabledDesc =
+            'Flag indicates whether or not ' +
+            this.props.category +
+            ' updates require explicit auditing approval process';
+        this.props.category === 'role' &&
+            this.props.justificationRequired &&
```
Comment (royelbaum) [inline ui/src/components/settings/SettingTable.js:374]: justificationRequired there is no need to pass it from father component can take it from redux

Comment (royelbaum) [reply ui/src/components/settings/SettingTable.js:374]: need to add || category === group

File: ui/src/components/settings/SettingTable.js [hunk 8]
```diff
@@ -355,6 +366,30 @@ class SettingTable extends React.Component {
...
+            rows.push(
+                <StyledSettingRow
+                    key={'setting-row-auditEnabled'}
+                    disabled={
+                        this.state.originalCollectionDetails.auditEnabled ||
+                        this.state.copyCollectionDetails.hasRoleMembers
```
Comment (royelbaum) [inline ui/src/components/settings/SettingTable.js:380]: need to add this.state.copyCollectionDetails.hasGroupMembers

**Review by royelbaum — CHANGES_REQUESTED** [review]
please add unit test for enabling and disabling the audit enable switch

File: ui/src/components/settings/SettingTable.js (hunk 8 above)
Comment (noy93845) [reply ui/src/components/settings/SettingTable.js]: fixed! apparently i didn't push the changes from vzm-ui, but now its synched.

File: ui/src/components/settings/SettingTable.js (hunk 7 above)
Comment (noy93845) [reply ui/src/components/settings/SettingTable.js]: removed justificationRequired and added the group category (apparently i didn't push the changes from vzm-ui, but now its synched.)

File: ui/src/components/role/AddRole.js (hunk 6 above)
Comment (noy93845) [reply ui/src/components/role/AddRole.js:160]: i cleaned it

File: ui/src/components/group/AddGroup.js (hunk 5 above)
Comment (noy93845) [reply ui/src/components/group/AddGroup.js:359]: i cleaned it

File: ui/src/components/settings/SettingRow.js [hunk 9]
```diff
@@ -103,6 +103,9 @@ export default class SettingRow extends React.Component {
                         value={this.props.value}
                         checked={this.props.value}
                         onChange={this.toggleSwitchButton}
+                        disabled={
+                            this.props.disabled ? this.props.disabled : false
+                        }
```
Comment (royelbaum) [inline ui/src/components/settings/SettingRow.js:108]: it look cleaner to write 
"this.props.disable || false"
has the same result

File: ui/src/__tests__/components/group/AddGroup.test.js [hunk 10]
```diff
@@ -390,4 +425,140 @@ describe('AddGroup', () => {
...
+        ).toHaveLength(1);
+
+        fireEvent.click(
+            screen.getByTestId('auditEnabled-switch-input')
+        );
+
+        await waitFor(() => {
+            expect(
+                screen.getAllByTitle(ADD_GROUP_AUDIT_ENABLED_TOOLTIP)
+            ).toHaveLength(2);
```
[the remark this answers is not in the document]
Comment (noy93845) [reply ui/src/__tests__/components/group/AddGroup.test.js:465]: if you look in AddGroup.js you will see that the `auditToolTip` is placed in two locations (the div for adding members and the members list)

File: ui/src/__tests__/components/group/AddGroup.test.js [hunk 11]
```diff
@@ -390,4 +425,140 @@ describe('AddGroup', () => {
...
+            getStateWithDomainData(domainData)
+        );
+
+        await waitFor(() =>
+            expect(getByTestId('add-modal-message')).toBeInTheDocument()
+        );
+
+        const addRoleForm = getByTestId('add-modal-message');
+        expect(addRoleForm).toMatchSnapshot();
+    });
```
[the remark this answers is not in the document]
Comment (noy93845) [reply ui/src/__tests__/components/group/AddGroup.test.js:558]: fixed and removed snapshot matching

File: ui/src/components/settings/SettingRow.js (hunk 9 above)
Comment (noy93845) [reply ui/src/components/settings/SettingRow.js]: done

**Review by royelbaum — APPROVED** [review]

**Review by havetisyan — APPROVED** [review]
